### PR TITLE
Windows 10 (64-bit) build instructions.  Disabled Lua script.  Restored config.h.in

### PR DESCRIPTION
**Pull request by brturn**
Lua script to load config file ~/.burrtools.rc was causing a seg-fault.  Haven't dug any deeper, just commented it out for now.

Bootstrap script was not creating the config.h.in file needed to complete configuration.  Restoring it for this push.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,4 +21,8 @@ src/burrTxt2
 src/gui/burrGui
 src/help/helpviewer
 src/unitTest
-src/config.h.in
+src/burrTxt.exe
```
Comment (akoehn) [inline .gitignore:24]: Why is config.h.in version-controlled again?  Is there a specific reason for not autogenerating it?

All the .exe files could be subsumed under `*.exe`

File: configure.ac [hunk 2]
```diff
@@ -20,7 +20,13 @@ AM_INIT_AUTOMAKE([subdir-objects])
 AM_SILENT_RULES([yes])
 AM_MAINTAINER_MODE
 
-AM_CONFIG_HEADER(src/config.h)
+AC_CONFIG_HEADERS([src/config.h])
+
+AC_CHECK_HEADER([winsock2.h],
```
Comment (akoehn) [inline configure.ac:25]: Is this something we have to guard to only be used in win32 environments? I'm not familiar with autotools+windows.

File: src/flu/Flu_Tree_Browser.cpp [hunk 3]
```diff
@@ -2983,7 +2983,7 @@ Flu_Tree_Browser::Node* Flu_Tree_Browser :: add_leaf( const char* path, const ch
 
 unsigned long Flu_Tree_Browser :: remove( const char *fullpath )
 {
-  return( (unsigned long)root.modify( fullpath, Node::REMOVE, rdata ) );
+  return reinterpret_cast<uintptr_t>(root.modify( fullpath, Node::REMOVE, rdata ) );
```
Comment (akoehn) [inline src/flu/Flu_Tree_Browser.cpp:2986]: This method returns a long, but the return is cast to uintptr.  Seems wrong to me.

File: src/flu/Flu_Tree_Browser.h [hunk 4]
```diff
@@ -1070,7 +1070,7 @@ class FLU_EXPORT Flu_Tree_Browser : public Fl_Group
       //! Remove the entry identified by path \b fullpath from this node
       /*! \return the unique id of the removed entry, or \c 0 if no matching entry was found */
       inline unsigned long remove( const char *fullpath )
-	{ return( (unsigned long)modify( fullpath, REMOVE, tree->rdata ) ); }
+	{ return reinterpret_cast<uintptr_t>(modify( fullpath, REMOVE, tree->rdata ) ); }
```
Comment (akoehn) [inline src/flu/Flu_Tree_Browser.h:1073]: Same as above.

File: src/lua/luaclass.cpp [hunk 5]
```diff
@@ -55,10 +55,16 @@ bool luaClass_c::getBool(const char *name) {
 
 /* functions to evaluate lua code */
 int luaClass_c::doFile(const char *fname) {
-  luaL_loadfile(L, fname) || lua_pcall(L, 0, 0, 0);
```
Comment (akoehn) [inline src/lua/luaclass.cpp:58]: I do not understand the reason to call lua_pcall.  Is this to extrract potential errors from the previous call?

**Review by akoehn — COMMENTED** [review]
See comments :-)

File: src/lua/luaclass.cpp (hunk 5 above)
Comment (akoehn) [reply src/lua/luaclass.cpp:58]: to add: Shouldn't the new code be 
```
  int ret = luaL_loadfile(L, fname);
  if (ret == 0) {
    return lua_pcall(L, 0, 0, 0);
  }
  return ret;
```

File: src/flu/Flu_Tree_Browser.cpp (hunk 3 above)
Comment (brturn) [reply src/flu/Flu_Tree_Browser.cpp:2986]: This is a 64-bit issue; the code is using a pointer as if it is a number (eg: pointer to unsigned long).  This is no longer valid in general (ptr size not matching size of a long).  The reinterpret cast changes the pointer value into a number (doing whatever is necessary for the platform at hand; uintptr is the magic type for this), then C++ auto-converts from that number to an unsigned long.

File: src/lua/luaclass.cpp (hunk 5 above)
Comment (brturn) [reply src/lua/luaclass.cpp:58]: This one is rather odd.  So the function returns an int, but has no return statement, which if I understand the semantics correctly means the "truthy" output of void, which is zero in integer space.  As for the Lua calls, the load file prepares the Lua VM for operation, loading and registering functions, etc.  The "pcall" actually executes the Lua VM.  Honestly don't care one way or another which return value to use.

File: .gitignore (hunk 1 above)
Comment (brturn) [reply .gitignore:24]: It's a duplicate line, it's still git-ignored.

File: configure.ac (hunk 2 above)
Comment (brturn) [reply configure.ac:25]: It works on my Mac and Linux build systems.  It is just checking for the location of a header, and if found, setting some environment variables.  In the case of non-windows systems, it does not find the header so the vars are empty.

File: src/flu/Flu_Tree_Browser.cpp (hunk 3 above)
Comment (akoehn) [reply src/flu/Flu_Tree_Browser.cpp:2986]: I understand the idea, but the code not casts long to uintptr and then -- if I understand it correctly -- the return value is cast to long again because the method returns long.  In the best case, this seems to be a noop.

On windows, size(long) < size(ptr), so there is a problem with addressability somewhere already and that underlying problem should be rather fixed.

In addition, this code does not compile on my debian with gcc 10 because `cstdint` is not included.

File: src/lua/luaclass.cpp (hunk 5 above)
Comment (akoehn) [reply src/lua/luaclass.cpp:58]: > So the function returns an int, but has no return statement, which if I understand the semantics correctly

According to the c++ standard, this is undefined behavior.  Therefore, I would return the lua return value.

File: src/flu/Flu_Tree_Browser.cpp (hunk 3 above)
Comment (brturn) [reply src/flu/Flu_Tree_Browser.cpp:2986]: Did some more digging.  The issue is that Node::modify returns a Node*, but the value is not always a pointer.  You can see this, for example around line 3353 the code returns the actual pointer to a Node, while in the same function, around line 3506 it returns the ID (an int) cast to a pointer.  The remove() function is documented in the header file to "return the unique ID of the node removed or 0 if none was found".  So, the reinterpret cast in this context is correct (pointer is converted to a number, zero is left alone).

Comment (akoehn) [reply src/flu/Flu_Tree_Browser.cpp:2986]: The point I was trying to make is that this method returns a long (line 2984) and therefore the cast inside the method is pointless because it is then cast to long again upon return

Comment (akoehn) [reply src/flu/Flu_Tree_Browser.cpp:2986]: Sorry its late here, will  reply with more sleep

Comment (brturn) [reply src/flu/Flu_Tree_Browser.cpp]: I couldn't remember why I used the reinterpret cast in the first place, so I checked in a hard-cast which compiles fine on my Mac.  Then I remembered that this is the Windows branch.. so back to Windows and MINGW... This is the error it is giving me with a hard-cast.  Reinterpret-cast compiles cleanly on all platforms for me (Mac, Ubuntu, Windows).

gui/../flu/Flu_Tree_Browser.h:1075:12: error: cast from 'Flu_Tree_Browser::Node*' to 'long unsigned int' loses precision [-fpermissive]
 1075 |     return (unsigned long)ret;
      |            ^~~~~~~~~~~~~~~~~~

Comment (akoehn) [reply src/flu/Flu_Tree_Browser.cpp]: Let me rephrase my critique: It does not matter what kind of casting is going on inside this method because the method itself returns a value of type double.  All the casts are doing is quieting the compiler.

Under Linux and Mac, long and pointer have the same size, so all the casts are noops. Under Win64, a long is only 32 bit but a pointer is 64 bit. The method returns a long, i.e. 32 bit. Converting a 32 bit double to uintptr (i.e. 64 bit) and then have the compiler add an implicit cast to double again (i.e. 32 bit) does not do anything meaningful as well.

The problem here are the return types that would need to be changed.

We can still merge this as it creates no new harm, but in that case we should have a TODO somewhere.
